clients/drcachesim/scheduler/flexible_queue.h

```diff
@@ -0,0 +1,219 @@
+    size() const
+    {
+        return entries_.size();
+    }
+
+    bool
+    find(T entry)
+    {
+        auto it = entry2index_.find(entry);
+        if (it == entry2index_.end())
+            return false;
+        return true;
+    }
+
+    bool
+    erase(T entry)
+    {
+        auto it = entry2index_.find(entry);
+        if (it == entry2index_.end())
+            return false;
+        index_t node = it->second;
+        if (node == entries_.size() - 1) {
+            entry2index_.erase(entry);
+            entries_.pop_back();
+            return true;
+        }
+        swap(node, entries_.size() - 1);
+        entry2index_.erase(entry);
+        entries_.pop_back();
+        percolate_down(node);
+        percolate_up(node);
+        print("after erase");
+        return true;
+    }
+
+private:
+    void
+    print(const std::string &message)
+    {
+        if (verbose_ <= 0)
+            return;
+        std::cout << message << "\n";
+        for (index_t i = 0; i < entries_.size(); ++i) {
+            std::cout << i << ": " << entries_[i] << " @ " << entry2index_[entries_[i]]
+                      << "\n";
+        }
+    }
+
+    inline index_t
+    parent_node(index_t node)
+    {
+        if (node <= 0)
+            return INVALID_INDEX;
+        return node / 2;
+    }
+    inline index_t
+    left_child(index_t node)
+    {
+        index_t child = node * 2;
+        if (child >= entries_.size())
+            return INVALID_INDEX;
+        return child;
+    }
+    inline index_t
+    right_child(index_t node)
+    {
+        index_t child = node * 2 + 1;
+        if (child >= entries_.size())
+            return INVALID_INDEX;
+        return child;
+    }
+    void
+    swap(index_t a, index_t b)
+    {
+        T temp = entries_[a];
+        entries_[a] = entries_[b];
+        entries_[b] = temp;
+        entry2index_[temp] = b;
+        entry2index_[entries_[a]] = a;
+    }
+    void
+    percolate_down(index_t node)
+    {
+        index_t should_be_parent = node;
+        index_t left = left_child(node);
+        index_t right = right_child(node);
+        if (left != INVALID_INDEX && !compare_(entries_[left], entries_[node]))
+            should_be_parent = left;
+        if (right != INVALID_INDEX && !compare_(entries_[right], entries_[node])) {
+            if (should_be_parent != left || !compare_(entries_[right], entries_[left]))
+                should_be_parent = right;
+        }
+        if (should_be_parent == node)
+            return;
+        swap(node, should_be_parent);
+        percolate_down(should_be_parent);
+    }
+
+    void
+    percolate_up(index_t node)
+    {
+        index_t parent = parent_node(node);
+        if (parent == INVALID_INDEX || !compare_(entries_[parent], entries_[node]))
+            return;
+        swap(node, parent);
+        percolate_up(parent);
+    }
+
+    std::vector<T> entries_;
+    comparator_t compare_;
+    std::unordered_map<T, index_t> entry2index_;
```

Comment:
Assumes T is trivially hashable. Ideally we should accept a hash function too in the template.

Reply:
Done.